### ISSUE TITLE
Reentrant versions of dask.utils.SerializableLock and distributed.Lock

### ISSUE THREAD

**Issue by shoyer** [dask/dask#3832]
I'm pretty sure to get distributed writes working properly in xarray (https://github.com/pydata/xarray/pull/2261) we will need reentrant versions of `dask.utils.SerializableLock` and `distributed.Lock`, i.e., which can be acquired multiple times by the same thread like [`threading.RLock`](https://docs.python.org/3.7/library/threading.html#rlock-objects). These could potentially be useful for other projects using dask as well.

Contemplating how to write these is hurting my head a little bit. In principle, we could just make a forked version of `SerializableLock` that uses `RLock` internally, but for `distributed.Lock` it's less clear.

Ideally, we could do this with composition, e.g., by writing `ReentrantLock(distributed.Lock(...))` to make a reentrant distributed lock. Does anyone have any idea what that would look like?

**Comment by mrocklin** [comment]
When someone calls acquire they assign some identifier to local state on the `ReentrantLock`, either thread-id or thread-id + worker address.  Subsequent calls to acquire/release increment/decrement some counter if the id is the same and operate on the underlying lock otherwise? 

It might also be simpler to just build a distributed.RLock.  This is a good task for someone who wants to understand how to extend dask.distribued.  It's somewhat accessible, lightly touches many parts of the system, and benefits from a few very similar examples in Lock, Variable, Queue, Pub/Sub, etc..

**Comment by shoyer** [comment]
> When someone calls acquire they assign some identifier to local state on the ReentrantLock, either thread-id or thread-id + worker address. Subsequent calls to acquire/release increment/decrement some counter if the id is the same and operate on the underlying lock otherwise?

Yes, this is roughly what I was thinking.

**Comment by jakirkham** [comment]
Would an RLock need to touch that much? Naively would think this would just affect the RLock object itself and not much else being as it is pretty simple extension to the Lock. Could RLock inherit from Lock?

**Comment by quasiben** [comment]
I think this issue is still valid but it looks like xarray is unblocked for now, correct ?

cc @shoyer

**Comment by shoyer** [comment]
Yes, we figured out another way. A reentrant lock would be a little cleaner but our solution works fine.

**Comment by jakirkham** [comment]
As we are revisiting this issue, am curious if anyone has thoughts on my comment above.
